nexus/src/app/background/tasks/support_bundle_collector.rs

```diff
@@ -513,40 +513,81 @@ impl BundleCollection {
+        //    blocked "await-ing" for this same underlying resource.
+        //
+        // In this specific case, the connection pool would be attempting to
+        // yield to the "&mut collection" arm, which cannot run, if we were
+        // blocking on the body of a different async select arm. This would
+        // result in a deadlock.
+        //
+        // In the future, we may attempt to make access to the connection pool
+        // safer from concurrent asynchronous access - it is unsettling that
+        // multiple concurrent ".claim()" functions can cause this behavior -
+        // but in the meantime, we spawn this cancellation check in an entirely
+        // new tokio task. Because of this separation, each task (the one
+        // checking for cancellation, and the main thread attempting to collect
+        // the bundle) do not risk preventing the other from being polled
+        // indefinitely.
+        let mut check_for_cancellation = tokio::task::spawn({
+            let s = self.clone();
+            async move {
+                loop {
+                    // Timer fired mid-collection - let's check if we should stop.
+                    yield_interval.tick().await;
                     trace!(
-                        &self.log,
+                        &s.log,
                         "Checking if Bundle Collection cancelled";
-                        "bundle" => %self.bundle.id
+                        "bundle" => %s.bundle.id
                     );
 
-                    let bundle = self.datastore.support_bundle_get(
-                        &self.opctx,
-                        self.bundle.id.into()
-                    ).await?;
+                    let bundle = s
+                        .datastore
+                        .support_bundle_get(&s.opctx, s.bundle.id.into())
+                        .await?;
                     if !matches!(bundle.state, SupportBundleState::Collecting) {
-                        warn!(
-                            &self.log,
-                            "Support Bundle cancelled - stopping collection";
-                            "bundle" => %self.bundle.id,
-                            "state" => ?self.bundle.state
-                        );
                         anyhow::bail!("Support Bundle Cancelled");
                     }
-                },
-                // Otherwise, keep making progress on the collection itself.
-                report = &mut collection => {
-                    info!(
-                        &self.log,
-                        "Bundle Collection completed";
-                        "bundle" => %self.bundle.id
-                    );
-                    return report;
-                },
+                }
             }
+        });
+
+        tokio::select! {
+            // Returns if the bundle has been cancelled explicitly, or if we
+            // cannot successfully check the bundle state.
```

Comment:
hmm, this makes me wonder if we actually want to have the cancellation checking task bail out if it encounters a DB error. might we want to just try again in another `yield_interval` seconds, in that case?

i suppose we really would want to handle different error types there, so that it bails out on e.g. `NotFound`s.